File: StandartTypesNative/TrainSingle.cpp
#define STANDARDTYPESAPI
#include "TrainSingle.h"

namespace StandardTypesNative {
	TrainSingle::~TrainSingle() {
		if (_inputData != 0) {
			delete [] _inputData;
			_inputLength = 0;
		}
	}

	TrainSingle::TrainSingle(float *inputData, int length) {
		_inputData = inputData;
		_inputLength = length;
	}

	TrainSingle::TrainSingle(const TrainSingle &source) {
		_inputLength = source._inputLength;
		float *sourceInputData = source._inputData;
		_inputData = new float[_inputLength];
		for(int i = 0; i < _inputLength; i++) {
			_inputData[i] = sourceInputData[i];
		}
	}

	float* TrainSingle::Input(void) const {
		return _inputData;
	}

	int TrainSingle::InputLength(void) const {
		return _inputLength;
	}
}

File: StandardTypes/TrainData/TrainSingle.cs
using System;
using System.Collections.Generic;

namespace StandardTypes {
	[Serializable]
	public class TrainSingle : TrainData, ICopyType<TrainSingle> {
		private float[] _inputData;
		private HashSet<int> _missedInputIndexes;

		public TrainSingle() {
		}

		public TrainSingle(float[] inputData, HashSet<int> missedInputIndexes = null) {
			_inputData = inputData;
			_missedInputIndexes = missedInputIndexes;
		}

		public TrainSingle(TrainSingle source) {
			var sourceInputData = source._inputData;
			_inputData = new float[sourceInputData.Length];
			sourceInputData.CopyTo(_inputData, 0);

			_missedInputIndexes = null;
			if (source._missedInputIndexes != null) {
				_missedInputIndexes = new HashSet<int>(source._missedInputIndexes);
			}

			Id = source.Id;
			Weight = source.Weight;
		}

		public float[] Input {
			get { return _inputData; }
			set { _inputData = value; }
		}

		public int InputLength {
			get { return _inputData.Length; }
		}

		public HashSet<int> MissedInputIndexes {
			get { return _missedInputIndexes; }
			set { _missedInputIndexes = value; }
		}

		public TrainSingle Copy() {
			return new TrainSingle(this);
		}
	}
}